syncer/ddl.go

```diff
@@ -295,8 +276,298 @@ func (s *Syncer) clearOnlineDDL(tctx *tcontext.Context, targetSchema, targetTabl
+		// for CREATE TABLE LIKE/AS, the reference tables should exist
+		shouldRefTableExistNum = len(srcTables)
+	case *ast.DropTableStmt:
+		shouldExecDDLOnSchemaTracker = true
+		if err := s.checkpoint.DeleteTablePoint(ec.tctx, srcTable.Schema, srcTable.Name); err != nil {
+			return err
+		}
+	case *ast.RenameTableStmt, *ast.CreateIndexStmt, *ast.DropIndexStmt, *ast.RepairTableStmt:
+		shouldExecDDLOnSchemaTracker = true
+		shouldSchemaExist = true
+		shouldTableExistNum = 1
+	case *ast.AlterTableStmt:
+		shouldSchemaExist = true
+		// for DDL that adds FK, since TiDB doesn't fully support it yet, we simply ignore execution of this DDL.
+		switch {
+		case len(node.Specs) == 1 && node.Specs[0].Constraint != nil && node.Specs[0].Constraint.Tp == ast.ConstraintForeignKey:
+			shouldTableExistNum = 1
+			shouldExecDDLOnSchemaTracker = false
+		case node.Specs[0].Tp == ast.AlterTableRenameTable:
+			shouldTableExistNum = 1
+			shouldExecDDLOnSchemaTracker = true
+		default:
+			shouldTableExistNum = len(srcTables)
+			shouldExecDDLOnSchemaTracker = true
+		}
+	case *ast.LockTablesStmt, *ast.UnlockTablesStmt, *ast.CleanupTableLockStmt, *ast.TruncateTableStmt:
+		break
+	default:
+		ec.tctx.L().DPanic("unhandled DDL type cannot be tracked", zap.Stringer("type", reflect.TypeOf(stmt)))
+	}
+
+	if shouldSchemaExist {
+		if err := s.schemaTracker.CreateSchemaIfNotExists(srcTable.Schema); err != nil {
+			return terror.ErrSchemaTrackerCannotCreateSchema.Delegate(err, srcTable.Schema)
+		}
+	}
+	for i := 0; i < shouldTableExistNum; i++ {
+		if _, err := s.getTable(ec.tctx, srcTables[i].Schema, srcTables[i].Name, targetTables[i].Schema, targetTables[i].Name); err != nil {
+			return err
+		}
+	}
+	// skip getTable before in above loop
+	// nolint:ifshort
+	start := 1
+	if shouldTableExistNum > start {
+		start = shouldTableExistNum
+	}
+	for i := start; i < shouldRefTableExistNum; i++ {
+		if err := s.schemaTracker.CreateSchemaIfNotExists(srcTables[i].Schema); err != nil {
+			return terror.ErrSchemaTrackerCannotCreateSchema.Delegate(err, srcTables[i].Schema)
+		}
+		if _, err := s.getTable(ec.tctx, srcTables[i].Schema, srcTables[i].Name, targetTables[i].Schema, targetTables[i].Name); err != nil {
+			return err
+		}
+	}
+
+	if shouldExecDDLOnSchemaTracker {
+		if err := s.schemaTracker.Exec(ec.tctx.Ctx, usedSchema, sql); err != nil {
+			ec.tctx.L().Error("cannot track DDL", zap.String("schema", usedSchema), zap.String("statement", sql), log.WrapStringerField("location", ec.currentLocation), log.ShortError(err))
+			return terror.ErrSchemaTrackerCannotExecDDL.Delegate(err, sql)
+		}
+		s.exprFilterGroup.ResetExprs(srcTable.Schema, srcTable.Name)
+	}
+	return nil
+}
+
+// filterAndResolveDDLSQL 1. parses the raw sql from upstream and returns the sqls that need to be handled. 2.handle onlineddl.
+func (s *Syncer) filterAndResolveDDLSQL(ev *replication.QueryEvent, ec eventContext, originSQL string) (sqls []string, tables map[string]*filter.Table, err error) {
```

Comment:
I think put this function in syncer.go will be better since it call other function like `skipBinlogDurationHistogram`